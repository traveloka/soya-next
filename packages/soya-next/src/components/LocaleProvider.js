import Router from "next/router";
import PropTypes from "prop-types";
import React from "react";
import { localeShape } from "../constants/PropTypes";
import LocaleContext from "../i18n/LocaleContext";

class LocaleProvider extends React.Component {
  static propTypes = {
    defaultLocale: PropTypes.string,
    siteLocales: PropTypes.arrayOf(PropTypes.string.isRequired),
    locale: localeShape,
    children: PropTypes.element.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
      defaultLocale: props.defaultLocale,
      siteLocales: props.siteLocales,
      locale: props.locale
    };
  }

  static getDerivedStateFromProps(props, state) {
    if (props.locale !== state.locale) {
      return { locale: props.locale };
    }
    return null;
  }

  render() {
    const { locale } = this.props;
    return (
      <LocaleContext.Provider value={this.state}>
        {React.Children.only(this.props.children)}
      </LocaleContext.Provider>
    );
  }
}

export default LocaleProvider;
