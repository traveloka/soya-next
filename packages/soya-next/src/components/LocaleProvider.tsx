import React from "react";
import LocaleContext from "../i18n/LocaleContext";
import type { SoyaNextLocaleContext } from "../types";

export interface LocaleProviderProps extends SoyaNextLocaleContext {
  children: React.ReactNode;
}
export interface LocaleProviderState extends SoyaNextLocaleContext {}

class LocaleProvider extends React.Component<
  LocaleProviderProps,
  LocaleProviderState
> {
  constructor(props: LocaleProviderProps) {
    super(props);
    this.state = {
      defaultLocale: props.defaultLocale,
      siteLocales: props.siteLocales,
      locale: props.locale,
    };
  }

  static getDerivedStateFromProps(
    props: LocaleProviderProps,
    state: LocaleProviderState
  ) {
    if (props.locale !== state.locale) {
      return { locale: props.locale };
    }
    return null;
  }

  render() {
    return (
      <LocaleContext.Provider value={this.state}>
        {React.Children.only(this.props.children)}
      </LocaleContext.Provider>
    );
  }
}

export default LocaleProvider;
