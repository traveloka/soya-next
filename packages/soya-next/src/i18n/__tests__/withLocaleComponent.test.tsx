import React from "react";
import { render } from "@testing-library/react";
import withLocale, {
  WithLocaleComponentInjectedProps,
} from "../withLocaleComponent";
import LocaleProvider from "../../components/LocaleProvider";

describe("withLocale", () => {
  function setupTest() {
    const context = {
      locale: {
        language: "id",
        country: "id",
      },
      defaultLocale: "id-id",
      siteLocales: ["id-id", "en-id"],
    };
    const Component = (props: WithLocaleComponentInjectedProps) => (
      <div data-testid={"component-locale-props"}>
        {JSON.stringify(props, null, 2)}
      </div>
    );
    const ComponentWithLocale = withLocale(Component);

    return { context, Component, ComponentWithLocale };
  }

  it("should add default locale, locale, and site locales to component props", () => {
    const { context, ComponentWithLocale } = setupTest();
    const { getByTestId } = render(
      <LocaleProvider {...context}>
        <ComponentWithLocale />
      </LocaleProvider>
    );
    const componentLocalePropsDiv = getByTestId("component-locale-props");
    expect(componentLocalePropsDiv.textContent).toMatchSnapshot();
  });

  it("should override locale context with locale props", () => {
    const { context, ComponentWithLocale } = setupTest();
    const locale = { language: "en", country: "id" };
    const { getByTestId } = render(
      <LocaleProvider {...context}>
        <ComponentWithLocale locale={locale} />
      </LocaleProvider>
    );
    const componentLocalePropsDiv = getByTestId("component-locale-props");
    expect(componentLocalePropsDiv.textContent).toMatchSnapshot();
  });

  it("should accept locale string props", () => {
    const { context, ComponentWithLocale } = setupTest();
    const { getByTestId } = render(
      <LocaleProvider {...context}>
        <ComponentWithLocale locale="en-id" />
      </LocaleProvider>
    );
    const componentLocalePropsDiv = getByTestId("component-locale-props");
    expect(componentLocalePropsDiv.textContent).toMatchSnapshot();
  });
});
