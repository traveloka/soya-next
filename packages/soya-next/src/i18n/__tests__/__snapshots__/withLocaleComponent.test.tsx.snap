// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withLocale should accept locale string props 1`] = `
"{
  \\"locale\\": {
    \\"country\\": \\"id\\",
    \\"language\\": \\"en\\"
  },
  \\"defaultLocale\\": \\"id-id\\",
  \\"siteLocales\\": [
    \\"id-id\\",
    \\"en-id\\"
  ]
}"
`;

exports[`withLocale should add default locale, locale, and site locales to component props 1`] = `
"{
  \\"defaultLocale\\": \\"id-id\\",
  \\"siteLocales\\": [
    \\"id-id\\",
    \\"en-id\\"
  ],
  \\"locale\\": {
    \\"language\\": \\"id\\",
    \\"country\\": \\"id\\"
  }
}"
`;

exports[`withLocale should override locale context with locale props 1`] = `
"{
  \\"locale\\": {
    \\"language\\": \\"en\\",
    \\"country\\": \\"id\\"
  },
  \\"defaultLocale\\": \\"id-id\\",
  \\"siteLocales\\": [
    \\"id-id\\",
    \\"en-id\\"
  ]
}"
`;
